lib/open_feature/hook.rb

```diff
@@ -2,5 +2,65 @@
 # frozen_string_literal: true
 
 module OpenFeature
-  class Hook < T::Struct; end
+  # See https://openfeature.dev/specification/sections/hooks
+  # We model Hooks as a simple ADT
+  module Hook
+    extend T::Sig
+    extend T::Helpers
+    interface!
+    sealed!
+
+    # See Requirement 4.3.2 - 4.3.4
+    class BeforeHook
```

Comment:
Interesting idea using a sealed class approach. I had thought that a `Hook` would have a `type` field set to an enum of the four available values. Then we'd have a `proc` field that can be called from the hook. What're your thoughts on that?

Reply:
This probably would have been my choice, except the sticky issue is that `BeforeHook.call` has a different function signature from the other ones. 

We could have a single Hook class with a default implementation that returns nil or something like that, but it seems a bit uglier to me.

Otherwise, it's the same in this case. wdyt?

Reply:
I _think_ the contracts could be the same. The requirement is that `before` hooks are the only ones allowed to modify the evaluation context. We could loosen the type to to allow a `T.nilable(EvaluationContext)` return and add some runtime checking to make sure only `before` contexts are editing contexts

Reply:
Thinking through this more, I think we don't want to require the different hook types to exist as their own classes. I think we want the one `Hook` class which accepts a `proc` that can be invoked. Then we include some housekeeping code to make sure only `before` hooks are allowed to alter evaluation contexts

Reply:
This works too! We wouldn't even need an additional check but can simply drop the eval context in other cases in the helper method. 

BUT It's a question of how much we want to lean on semantics of sorbet vs runtime implementation. In general, I favor the static side -- in this case, using the sealed capability to define a little algebraic data type -- because it is a formal description. At scale this has huge advantages. Here the *only* salient difference is whether we will define the functional signature of the hook `call` method accurately or rely on users to understand the spec.